package/utils/px5g-wolfssl/px5g-wolfssl.c

```diff
@@ -0,0 +1,301 @@
+#include <wolfssl/wolfcrypt/asn.h>
+#include <wolfssl/wolfcrypt/asn_public.h>
+#include <wolfssl/wolfcrypt/ecc.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/rsa.h>
+#include <wolfssl/wolfcrypt/settings.h>
+
+#define HEAP_HINT NULL
+#define FOURK_SZ 4096
+
+enum {
+    EC_KEY_TYPE = 0,
+    RSA_KEY_TYPE = 1,
+};
+
+int write_file(byte* buf, int bufSz, char* path)
+{
+    int ret;
+    FILE* file;
+    if (path) {
+	file = fopen(path, "wb");
+    } else {
+	file = stdout;
+    }
+    ret = (int)fwrite(buf, 1, bufSz, file);
+    if (path) {
+	fclose(file);
+    }
+    if (ret > 0) {
+	/* ret > 0 indicates a successful file write, set to zero for return */
+	ret = 0;
+    }
+    return ret;
+}
+
+int write_key(ecc_key* ecKey, RsaKey* rsaKey, int type, int keySz, char* fName, bool write_pem)
+{
+    int ret;
+    byte der[FOURK_SZ], pem[FOURK_SZ];
+    int derSz, pemSz;
+    XMEMSET(der, 0, FOURK_SZ);
+    XMEMSET(pem, 0, FOURK_SZ);
+    if (type == EC_KEY_TYPE) {
+	ret = wc_EccKeyToDer(ecKey, der, sizeof(der));
+    } else {
+	ret = wc_RsaKeyToDer(rsaKey, der, sizeof(der));
+    }
+    if (ret <= 0) {
+	printf("Key To DER failed: %d\n", ret);
+    }
+    derSz = ret;
+
+    if (write_pem) {
+	memset(pem, 0, sizeof(pem));
+	if (type == EC_KEY_TYPE) {
+	    ret = wc_DerToPem(der, derSz, pem, sizeof(pem), ECC_PRIVATEKEY_TYPE);
+	} else {
+	    ret = wc_DerToPem(der, derSz, pem, sizeof(pem), PRIVATEKEY_TYPE);
+	}
+	if (ret <= 0) {
+	    printf("DER to PEM failed: %d\n", ret);
+	}
+	pemSz = ret;
+	ret = write_file(pem, pemSz, fName);
+    } else {
+	ret = write_file(der, derSz, fName);
+    }
+    return ret;
+}
+
+int gen_key(WC_RNG* rng, ecc_key* ecKey, RsaKey* rsaKey, int type, int keySz, long exp)
+{
+    int ret;
+
+    if (type == EC_KEY_TYPE) {
+	ret = wc_ecc_init(ecKey);
+	(void)rsaKey; /* Not using rsaKey in EC_KEY_TYPE case */
+    } else {
+	ret = wc_InitRsaKey(rsaKey, NULL);
+	(void)ecKey; /* Not using ecKey in RSA_KEY_TYPE case */
+    }
+    if (ret != 0) {
+	printf("Key initialization failed: %d\n", ret);
+    }
+
+    if (type == EC_KEY_TYPE) {
+	fprintf(stderr, "Generating EC private key\n");
+	ret = wc_ecc_make_key_ex(rng, keySz, ecKey, ECC_SECP256R1);
+    } else {
+	fprintf(stderr, "Generating RSA private key, %i bit long modulus\n", keySz);
+	ret = wc_MakeRsaKey(rsaKey, keySz, WC_RSA_EXPONENT, rng);
+    }
+    if (ret != 0) {
+	printf("Key generation failed: %d\n", ret);
+    }
+    return ret;
+}
+
+int selfsigned(WC_RNG* rng, char** arg)
+{
+    ecc_key ecKey;
+    RsaKey rsaKey;
+    int ret;
+    char* subject = "";
+    int keySz = 2048;
+    int type = RSA_KEY_TYPE;
+    int exp = 65537;
+    unsigned int days = 30;
+    char *keypath = NULL, *certpath = NULL;
+    char fstr[20], tstr[20];
+    bool pem = true;
+    Cert newCert;
+    time_t from = time(NULL), to;
+    byte derBuf[FOURK_SZ];
+    byte pemBuf[FOURK_SZ];
+    int pemSz = -1;
+    int derSz = -1;
+    wc_InitCert(&newCert);
```

Comment:
Missing error handling.

Reply:
Actually this was redundant with another init a few lines later.